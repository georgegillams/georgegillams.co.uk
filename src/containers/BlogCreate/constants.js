export const KEY = 'blog-list';
export const CREATE_BLOG = `app/${KEY}/CREATE_BLOG`;
