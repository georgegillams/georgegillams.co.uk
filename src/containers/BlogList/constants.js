export const KEY = 'blog-list';
export const LOAD_BLOGS = `app/${KEY}/LOAD_BLOGS`;
export const DELETE_BLOG = `app/${KEY}/DELETE_BLOG`;
