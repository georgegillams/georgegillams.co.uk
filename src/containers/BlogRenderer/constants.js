export const KEY = 'blog-renderer';
export const LOAD_BLOG = `app/${KEY}/LOAD_BLOG`;
export const UPDATE_BLOG = `app/${KEY}/UPDATE_BLOG`;
