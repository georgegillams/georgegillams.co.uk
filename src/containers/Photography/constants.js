export const KEY = 'photography';
export const LOAD_PHOTOS = `app/${KEY}/LOAD_PHOTOS`;
