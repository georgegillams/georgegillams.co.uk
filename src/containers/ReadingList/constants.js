export const KEY = 'reading-list';
export const LOAD_BOOKS = `app/${KEY}/LOAD_BOOKS`;
export const UPDATE_BOOK = `app/${KEY}/UPDATE_BOOK`;
export const DELETE_BOOK = `app/${KEY}/DELETE_BOOK`;
