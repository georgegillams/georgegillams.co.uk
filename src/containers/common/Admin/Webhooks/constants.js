export const KEY = 'webhooks';

export const LOAD_ENDPOINTS = `app/${KEY}/LOAD_ENDPOINTS`;
export const CREATE_ENDPOINT = `app/${KEY}/CREATE_ENDPOINT`;
export const UPDATE_ENDPOINT = `app/${KEY}/UPDATE_ENDPOINT`;
export const REMOVE_ENDPOINT = `app/${KEY}/REMOVE_ENDPOINT `;

export const LOAD_NOTIFICATIONS = `app/${KEY}/LOAD_NOTIFICATIONS`;
export const REMOVE_NOTIFICATION = `app/${KEY}/REMOVE_NOTIFICATION`;
