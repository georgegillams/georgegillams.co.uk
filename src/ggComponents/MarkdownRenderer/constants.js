const DEFAULT_SUPPORTED_FEATURES = [
  'code',
  'blockCode',
  'bold',
  'italic',
  'strikethrough',
  'quotation',
  'bigLink',
  'link',
  'footnote',
  'image',
  'video',
  'citation',
  'sectioning',
  'linebreak',
  'bulletList',
  'numberedList',
];

export default DEFAULT_SUPPORTED_FEATURES;
export { DEFAULT_SUPPORTED_FEATURES };
