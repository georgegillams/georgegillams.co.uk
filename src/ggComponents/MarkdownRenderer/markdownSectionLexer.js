/* eslint-disable no-useless-escape */
const MD_SECTION_REGEX = /^#\ (.+)/gims;
const MD_SUBSECTION_REGEX = /^##\ (.+)/gims;
const MD_SUBSUBSECTION_REGEX = /^###\ (.+)/gims;
/* eslint-enable */

const matchesSectioningRegex = text => {
  let matches = false;
  if (text.match(MD_SECTION_REGEX)) {
    matches = true;
  }
  if (text.match(MD_SUBSECTION_REGEX)) {
    matches = true;
  }
  if (text.match(MD_SUBSUBSECTION_REGEX)) {
    matches = true;
  }
  return matches;
};

const isSectioned = splitContent => {
  let contentIsSectioned = false;
  splitContent.forEach(c => {
    if (matchesSectioningRegex(c)) {
      contentIsSectioned = true;
    }
  });
  return contentIsSectioned;
};

const convertStringsToLines = strings => strings.map(sC => ({ type: 'line', content: sC }));

// TODO Should ignore lines matching the regex if inside a code block
/*
 * Turns a flat list of strings into a list of sections (and paragraphs), extracting the name and children for each one.
 */
const parseSectioning = (splitContent, type, regex, listIfUnsectioned = true) => {
  if (splitContent.length === 0) {
    return { type, name: null, children: [] };
  }

  const results = [];

  let currentResult = null;
  if (!listIfUnsectioned) {
    currentResult = { type, name: null, children: [] };
  }

  splitContent.forEach(sC => {
    if (sC.match(regex)) {
      results.push(currentResult);
      const title = sC.split(regex)[1];
      currentResult = { type, name: title, children: [] };
    } else if (currentResult) {
      currentResult.children.push(sC);
    } else {
      results.push({ type: 'line', content: sC });
    }
  });

  results.push(currentResult);
  const nonEmptyResults = results.filter(x => x && (x.type !== type || !!x.name || x.children.length > 0));

  return nonEmptyResults;
};

const parseSubsubsections = splitContent => {
  const result = parseSectioning(splitContent, 'subsubsection', MD_SUBSUBSECTION_REGEX);
  return result;
};

const parseSubsections = splitContent => {
  let result = parseSectioning(splitContent, 'subsection', MD_SUBSECTION_REGEX);
  result = result.map(r => {
    if (r.type === 'subsection') {
      return {
        ...r,
        children: parseSubsubsections(r.children),
      };
    }
    return r;
  });
  return result;
};

const parseSections = splitContent => {
  let result = parseSectioning(splitContent, 'section', MD_SECTION_REGEX, false);
  result = result.map(r => {
    if (r.type === 'section') {
      return { ...r, children: parseSubsections(r.children) };
    }
    return r;
  });
  return result;
};

/*
 * Turns a string into a list of sections and lines, ready for further processing
 */
const lexStructure = (content, supportedFeatures) => {
  const splitContent = content.split('\n').filter(x => x.trim() !== '');
  if (splitContent.length === 0) {
    return [];
  }
  if (!supportedFeatures.includes('sectioning') || !isSectioned(splitContent)) {
    return convertStringsToLines(splitContent);
  }
  const parsedSections = parseSections(splitContent);
  return parsedSections;
};

export default lexStructure;
export { lexStructure };
