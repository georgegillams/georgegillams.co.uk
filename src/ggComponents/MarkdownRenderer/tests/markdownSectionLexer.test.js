/* eslint-disable max-len */

import { lexStructure } from '../markdownSectionLexer.js';
import { DEFAULT_SUPPORTED_FEATURES } from '../constants';

test('parses whitespace', () => {
  const result = lexStructure('  \n \n   ', DEFAULT_SUPPORTED_FEATURES);
  expect(result.error).toBe(undefined);
  expect(result).toMatchSnapshot();
});

test('parses single paragraph', () => {
  const result = lexStructure('1 paragraph.', DEFAULT_SUPPORTED_FEATURES);
  expect(result.error).toBe(undefined);
  expect(result).toMatchSnapshot();
});

test('parses multiple paragraphs', () => {
  const result = lexStructure('1 paragraph.\n2 paragraph.\n3 paragraph.', DEFAULT_SUPPORTED_FEATURES);
  expect(result.error).toBe(undefined);
  expect(result).toMatchSnapshot();
});

test("doesn't parses section if unsupported", () => {
  const result = lexStructure('# 1 Unsupported section\n1 paragraph.', []);
  expect(result.error).toBe(undefined);
  expect(result).toMatchSnapshot();
});

test('parses section at start', () => {
  const result = lexStructure('# 1 Section\n1 paragraph.', DEFAULT_SUPPORTED_FEATURES);
  expect(result.error).toBe(undefined);
  expect(result).toMatchSnapshot();
});

test('parses section in middle', () => {
  const result = lexStructure('1 paragraph.\n# 1 Section\n2 paragraph.', DEFAULT_SUPPORTED_FEATURES);
  expect(result.error).toBe(undefined);
  expect(result).toMatchSnapshot();
});

test('parses nested sectioning', () => {
  const result = lexStructure(
    '1 paragraph.\n# 1 Section\n2 paragraph.\n## 1 Sub-section\n3 paragraph.\n# 2 Sub-section\n4 paragraph.\n# 2 Section\n5 paragraph.',
    DEFAULT_SUPPORTED_FEATURES
  );
  expect(result.error).toBe(undefined);
  expect(result).toMatchSnapshot();
});
