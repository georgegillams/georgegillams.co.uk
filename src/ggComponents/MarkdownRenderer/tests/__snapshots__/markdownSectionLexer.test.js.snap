// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doesn't parses section if unsupported 1`] = `
Array [
  Object {
    "content": "# 1 Unsupported section",
    "type": "line",
  },
  Object {
    "content": "1 paragraph.",
    "type": "line",
  },
]
`;

exports[`parses multiple paragraphs 1`] = `
Array [
  Object {
    "content": "1 paragraph.",
    "type": "line",
  },
  Object {
    "content": "2 paragraph.",
    "type": "line",
  },
  Object {
    "content": "3 paragraph.",
    "type": "line",
  },
]
`;

exports[`parses nested sectioning 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "content": "1 paragraph.",
        "type": "line",
      },
    ],
    "name": null,
    "type": "section",
  },
  Object {
    "children": Array [
      Object {
        "content": "2 paragraph.",
        "type": "line",
      },
      Object {
        "children": Array [
          Object {
            "content": "3 paragraph.",
            "type": "line",
          },
        ],
        "name": "1 Sub-section",
        "type": "subsection",
      },
    ],
    "name": "1 Section",
    "type": "section",
  },
  Object {
    "children": Array [
      Object {
        "content": "4 paragraph.",
        "type": "line",
      },
    ],
    "name": "2 Sub-section",
    "type": "section",
  },
  Object {
    "children": Array [
      Object {
        "content": "5 paragraph.",
        "type": "line",
      },
    ],
    "name": "2 Section",
    "type": "section",
  },
]
`;

exports[`parses section at start 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "content": "1 paragraph.",
        "type": "line",
      },
    ],
    "name": "1 Section",
    "type": "section",
  },
]
`;

exports[`parses section in middle 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "content": "1 paragraph.",
        "type": "line",
      },
    ],
    "name": null,
    "type": "section",
  },
  Object {
    "children": Array [
      Object {
        "content": "2 paragraph.",
        "type": "line",
      },
    ],
    "name": "1 Section",
    "type": "section",
  },
]
`;

exports[`parses single paragraph 1`] = `
Array [
  Object {
    "content": "1 paragraph.",
    "type": "line",
  },
]
`;

exports[`parses whitespace 1`] = `Array []`;
